### DISCUSSION TITLE
Why is overriding initWithLayer necessary in video layer?

### DISCUSSION THREAD

**Discussion by krackers** [iina/iina#5046] [General]
The `  override init(layer: Any) {` was added in https://github.com/iina/iina/commit/33c585f78f604dd59d056db4cd836e4d6cafc976 apparently to fix something with multi-screen behavior. Mpv has something similar, mentioning `    // necessary for when the layer containing window changes the screen`

@lhc70000 Do you remember why exactly this is needed? I tried but could not get this to ever fire, and the apple documentation implies it is only ever called when you animate a layer.

**Comment by krackers** [comment]
I ask because IINA's current implementation is not strictly correct, the shadow layer created could (will? not sure tbh. But I'd expect it to be separate since they have different FBO rendering targets)  have its own CGLContext, and if mpv tries to render using it it won't be able to access any of the textures previously allocated. Mpv's implementation is more correct in that it copies over the cglcontext as well. But having the possibility of a shadow layer makes thinking about all the edge-cases even harder than it already is, if there is no situation in which this is called it'd seem to be safer to just `fatalError`.

**Comment by low-batt** [comment]
I too wondered why this constructor was needed. The mpv code has this comment:
>     // necessary for when the layer containing window changes the screen

I probably need to update PR #5029 to copy over the OpenGL context like mpv does.

I will experiment with multiple monitors tomorrow.

**Reply by krackers** [reply]
The mpv code was written after IINA's code, and I suspect IINA might have been used as a reference or something. Both also mistakenly call `super.init` instead of `super.initWithLayer`.

@lhc70000 might hopefully remember why this was added. But the fact that this has been broken (doesn't copy over CGL Context) and yet no one has complained about anything makes me think it is unnecessary, and it would be safer to just stub it out similar to `initWithCoder`.

Btw mpv is wrong in another way, it does not copy over the `displayLock`. But it doesn't matter since CAOpenGLLayer, like most of CoreAnimation(?) is fairly thread-safe from what I can see

**Comment by lhc70000** [comment]
From the commit, apparently it was trying to apply the correct scale when two screens have different DPI scaling.

However, from my memory, this `override init()` wasn't the key to solve this issue. It could be a byproduct of trial and error. It's also possible that on OSX 10.11, under some rare cases, the system will call this method when you move the layer to another screen with different scaling.

> the shadow layer created could (will? not sure tbh. But I'd expect it to be separate since they have different FBO rendering targets) have its own CGLContext

If this is true AND the system did call this initializer, there would be a crash or at least a critical problem that will be reported by users. However we didn't receive such report. So yes, I think there's very little chance it can be invoked, especially on modern macOS versions.

> The mpv code was written after IINA's code, and I suspect IINA might have been used as a reference or something.

Yes, that happened maybe in 2018 or 2019.

> it'd seem to be safer to just fatalError.

I'd like to put it in the 1.4.0 beta 1, but we already have so many features in this version and I don't want to introduce a potential crash. Perhaps we can copy the OpenGL context like mpv for now, and do this experiment in the next beta version?

**Comment by low-batt** [comment]
I can add copying the context like mpv does to PR #5029. 

I can also add a log message in this constructor. If we get a log file from a user with that message in it we will know for certain it is being used.

Both IINA and mpv are definitely not following this guidance from the [init(layer:)](https://developer.apple.com/documentation/quartzcore/calayer/1410842-init) documentation:
> Subclasses should always invoke the superclass implementation.

But that does say "should" and not "must", so maybe allowed?

**Comment by low-batt** [comment]
The Apple documentation for [init(layer:)](https://developer.apple.com/documentation/quartzcore/calayer/1410842-init) says:
> This initializer is used to create shadow copies of layers, for example, for the [presentation()](https://developer.apple.com/documentation/quartzcore/calayer/1410744-presentation) method.

So as a test I added some test code that called `presentation()` and added a breakpoint in `init(layer:)` to check that the constructor is being called and doesn't trigger a crash.

The results were a surprise to me. Calling `presentation()` did not invoke the `init(layer:)` constructor. I'm puzzled by that result.

**Comment by krackers** [comment]
@low-batt It seems that this method can be called when `-[CALayer setContentScale]` is called. I don't want to mess with my display settings too much to test (I have hiDPI enabled on both internal retina and external display) but I found that with only the external display active, unplugging the display will trigger a backing properties change and cause crash if `initWithLayer` is stubbed out:

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libswiftCore.dylib            	       0x19628dee0 _assertionFailure(_:_:file:line:flags:) + 268
1   mpv                           	       0x104952020 VideoLayer.init(layer:) + 644
2   mpv                           	       0x104952060 @objc VideoLayer.init(layer:) + 56
3   QuartzCore                    	       0x18e894904 CA::Layer::presentation_layer(CA::Transaction*) + 348
4   QuartzCore                    	       0x18e8a374c CALayerCreateImplicitAnimation + 140
5   QuartzCore                    	       0x18e865cf0 -[CALayer actionForKey:] + 420
6   QuartzCore                    	       0x18e860b88 CA::Layer::begin_change(CA::Transaction*, unsigned int, objc_object*, objc_object*&) + 204
7   QuartzCore                    	       0x18e867938 -[CALayer setContentsScale:] + 212
8   mpv                           	       0x104955da4 Window.windowDidChangeBackingProperties(_:) + 108
9   mpv                           	       0x104956450 @objc Window.windowDidEnterFullScreen(_:) + 108
10  CoreFoundation                	       0x18665f144 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 148
11  CoreFoundation                	       0x1866f33d8 ___CFXRegistrationPost_block_invoke + 88
12  CoreFoundation                	       0x1866f3320 _CFXRegistrationPost + 440
13  CoreFoundation                	       0x18662d678 _CFXNotificationPost + 768
14  Foundation                    	       0x18774a4e4 -[NSNotificationCenter postNotificationName:object:userInfo:] + 88
15  AppKit                        	       0x18a981318 -[NSWindow _postWindowDidChangeBackingPropertiesAndDisplayWindowForPreviousBackingScaleFactor:previousColorSpace:] + 164
16  AppKit                        	       0x18a08da54 __67-[NSWindow _updateSettingsSendingScreenChangeNotificationToScreen:]_block_invoke + 448
17  AppKit                        	       0x189f0a8a4 NSPerformVisuallyAtomicChange + 108
18  AppKit                        	       0x18a08d888 -[NSWindow _updateSettingsSendingScreenChangeNotificationToScreen:] + 96
19  AppKit                        	       0x18a08dc5c -[NSWindow _updateSettingsSendingScreenChangeNotification] + 128
```

(Note that I think the presence of `windowDidEnterFullScreen` is misleading and possibly due to an incorrectly parsed backtrace, I set a breakpoint at the beginning of that frame and it was never hit.  The video was also not in fullscreen. I do not have much experience in swift debugging so I don't know why it shows up.)

So it does seem that when the scale factor changes, CoreAnimation tries to do something with a shadow-layer (presumably it displays the shadow-layer until the underlying actual layer redraws at the new resolution or something). Btw I'm too lazy to debug, but if you break on `-[CALayer actionForKey:]` I'm curious what the key it's trying to animate is... I'm guessing it's `contentScale`.

**Comment by low-batt** [comment]
@krackers Reproduced for me with `init(layer:)` commented out:
![crash](https://github.com/user-attachments/assets/9fd0e5c5-471d-4cc6-b642-6d46f764853f)

Now that we know how to trigger the use of this constructor, the next question is how to trigger the suspected bad behavior of IINA's faulty implementation.

**Reply by krackers** [reply]
>trigger the suspected bad behavior of IINA's faulty implementation.

I suppose you'd have to see if `draw` is even called on the shadow layer that's created (or if the layer is just created but not actually used). You could maybe log the address of `self` on all draw invocations and see if it matches with the address of `self` in `initWithLayer`

**Reply by low-batt** [reply]
I will try that to do that. Currently battling two other problems. I consider this one a high priority too.

**Comment by low-batt** [comment]
@krackers Please have a look at PR #5086.
